### PR TITLE
8303882: Refactor some iterators in jdk.compiler

### PR DESCRIPTION
**Pull request by pavelrappo**
Please review this refactoring to iterators in jdk.compiler. The refactoring delegates more to collections framework, while retaining the performance characteristics of bespoke iterators. The refactoring also adds inline comments, `@Override` annotations, and fixes some trivial bugs (as can be seen in the added test, which is slightly augmented from that suggested by Jan Lahoda here https://github.com/openjdk/jdk/pull/12904#discussion_r1129777660).

I'll add a comment here once I have benchmarked the change formally.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8303882](https://bugs.openjdk.org/browse/JDK-8303882): Refactor some iterators in jdk.compiler

### Reviewers
 * [Maurizio Cimadamore](https://openjdk.org/census#mcimadamore) (@mcimadamore - **Reviewer**) ⚠️ Review applies to [04596f39](https://git.openjdk.org/jdk/pull/12949/files/04596f393c73426a9a48f750365bdff16402bacc)

### Contributors
 * Jan Lahoda `<jlahoda@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12949/head:pull/12949` \
`$ git checkout pull/12949`

Update a local copy of the PR: \
`$ git checkout pull/12949` \
`$ git pull https://git.openjdk.org/jdk pull/12949/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12949`

View PR using the GUI difftool: \
`$ git pr show -t 12949`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12949.diff">https://git.openjdk.org/jdk/pull/12949.diff</a>

</details>

### CODE REVIEW

File: test/langtools/tools/javac/util/IteratorsTest.java [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+        }
+
+        @Override
+        public T next() {
+            nextCalls++;
+            return delegate.next();
+        }
+    }
+
+    static class TestConverter<I, O>  implements Function<I, Iterator<O>> {
```
Comment (turbanoff) [inline test/langtools/tools/javac/util/IteratorsTest.java:150]:
```suggestion
    static class TestConverter<I, O> implements Function<I, Iterator<O>> {
```

**Comment by pavelrappo** [comment]
I've run a relevant performance test suit on this PR: the results were noisy and inconclusive. That said, there was no gross degradation. I think it's time to make a choice: either let this PR in or drop it because it has already burned too much review cycles.

From my perspective, this PR improves code reuse while also retaining characteristics of the compound iterator: the iterator is still lazy with the minimum number of calls to constituent iterators.

File: test/langtools/tools/javac/util/IteratorsTest.java [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+            this.delegate = delegate;
+        }
+
+        @Override
+        public Iterator<O> apply(I t) {
+            applyCalls++;
+            return delegate.apply(t);
+        }
+    }
+}
```
Comment (mcimadamore) [inline test/langtools/tools/javac/util/IteratorsTest.java:172]: Missing newline here

File: test/langtools/tools/javac/util/IteratorsTest.java [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+            return delegate.next();
+        }
+    }
+
+    static class TestConverter<I, O> implements Function<I, Iterator<O>> {
+        int applyCalls;
+        final Function<I, Iterator<O>> delegate;
+
+        public TestConverter(Function<I, Iterator<O>> delegate) {
+            this.delegate = delegate;
```
Comment (mcimadamore) [inline test/langtools/tools/javac/util/IteratorsTest.java:163]: This parameter doesn't really seem to be used in the test?

Comment (mcimadamore) [reply test/langtools/tools/javac/util/IteratorsTest.java:163]: I suggest this TestConverter to take a single type parameter `T` and implement `Function<TestIterator<S>, Iterator<S>>`

File: test/langtools/tools/javac/util/IteratorsTest.java [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+        assertAndResetMaxCalls(c, 0);
+        assertAndResetMaxCalls(test1, 1, 0);
+        assertAndResetMaxCalls(test2, 0, 0);
+
+        //next may invoke hasNext once:
+        Assertions.assertEquals("1", compound.next());
+
+        assertAndResetMaxCalls(c, 0);
+        assertAndResetMaxCalls(test1, 1, 1);
+        assertAndResetMaxCalls(test2, 0, 0);
```
Comment (mcimadamore) [inline test/langtools/tools/javac/util/IteratorsTest.java:112]: Should the test continue, and check that we end up with calls to the second iterator?

**Review by mcimadamore — APPROVED** [review]
Looks good - some minor (optional) comments on the test

**Comment by pavelrappo** [comment]
@mcimadamore, I have updated the test. Please check if that's what you had in mind.

**Pull request closed without merging**